pkg/kubectl/cmd/version.go

```diff
@@ -50,73 +58,90 @@ func NewCmdVersion(f cmdutil.Factory, out io.Writer) *cobra.Command {
 		Long:    "Print the client and server version information for the current context",
 		Example: versionExample,
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunVersion(f, out, cmd)
-			cmdutil.CheckErr(err)
+			options := new(VersionConfig)
+			cmdutil.CheckErr(options.Complete(cmd))
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run(f, out))
 		},
 	}
 	cmd.Flags().BoolP("client", "c", false, "Client version only (no server required).")
 	cmd.Flags().BoolP("short", "", false, "Print just the version number.")
-	cmd.Flags().String("output", "", "output format, options available are yaml and json")
+	cmd.Flags().String("output", "", "one of 'yaml', 'json', or ''")
 	cmd.Flags().MarkShorthandDeprecated("client", "please use --client instead.")
 	return cmd
 }
 
-func RunVersion(f cmdutil.Factory, out io.Writer, cmd *cobra.Command) error {
-	var serverVersion *apimachineryversion.Info = nil
-	var serverErr error = nil
-	vo := Version{nil, nil}
+func retrieveServerVersion(f cmdutil.Factory) (*apimachineryversion.Info, error) {
+	discoveryClient, err := f.DiscoveryClient()
+	if err != nil {
+		return nil, err
+	}
+
+	// Always request fresh data from the server
+	discoveryClient.Invalidate()
+	return discoveryClient.ServerVersion()
+}
+
+func (v *VersionConfig) Run(f cmdutil.Factory, out io.Writer) error {
+	var (
+		serverVersion *apimachineryversion.Info
+		serverErr     error
+		versionInfo   Version
+	)
 
 	clientVersion := version.Get()
-	vo.ClientVersion = &clientVersion
+	versionInfo.ClientVersion = &clientVersion
 
-	if !cmdutil.GetFlagBool(cmd, "client") {
+	if !v.clientOnly {
 		serverVersion, serverErr = retrieveServerVersion(f)
-		vo.ServerVersion = serverVersion
+		versionInfo.ServerVersion = serverVersion
 	}
 
-	switch of := cmdutil.GetFlagString(cmd, "output"); of {
-	case "":
-		if cmdutil.GetFlagBool(cmd, "short") {
+	if v.output == "" {
+		if v.short {
 			fmt.Fprintf(out, "Client Version: %s\n", clientVersion.GitVersion)
-
 			if serverVersion != nil {
 				fmt.Fprintf(out, "Server Version: %s\n", serverVersion.GitVersion)
 			}
 		} else {
 			fmt.Fprintf(out, "Client Version: %s\n", fmt.Sprintf("%#v", clientVersion))
-
 			if serverVersion != nil {
 				fmt.Fprintf(out, "Server Version: %s\n", fmt.Sprintf("%#v", *serverVersion))
 			}
 		}
-	case "yaml":
-		y, err := yaml.Marshal(&vo)
-		if err != nil {
-			return err
+	} else {
+		var marshalled []byte
+		var err error
+		if v.output == "yaml" {
+			marshalled, err = yaml.Marshal(&versionInfo)
+		} else if v.output == "json" {
+			marshalled, err = json.Marshal(&versionInfo)
+		} else {
+			panic(fmt.Sprintf(
```

Comment:
This is only possible to hit if there is a bug in the program.
I don't think it makes sense to return an error for that situation.
What do you think?